### PR TITLE
Fixed #26808 -- Introduce Meta.indexes

### PR DESCRIPTION
**Pull request by akki**
Ticket [#26808](https://code.djangoproject.com/ticket/26808).

Other PR that needs to be merged before this - [~~!6871~~](https://github.com/django/django/pull/6871).

### CODE REVIEW

**Comment by timgraham** [comment]
As long as the subsequent commits don't rely on any of the previous ones, it's certainly easier to review smaller pull requests.

File: docs/ref/schema-editor.txt [hunk 1]
```diff
@@ -67,6 +67,20 @@ unique constraints or indexes it requires.
 Drops the model's table in the database along with any unique constraints
 or indexes it has.
 
+``add_index()``
+---------------
+
+.. method:: BaseDatabaseSchemaEditor.add_index(model, index)
+
```
Comment (timgraham) [inline docs/ref/schema-editor.txt]: .. versionadded:: 1.11

**Comment by akki** [comment]
But I am worried that commits like this https://github.com/django/django/pull/6857/commits/0e2cdc2500d7e7e39f551e1455d41d86567f4cfa might look of no use unless seen in context of the subsequent commits.

**Comment by timgraham** [comment]
You can add "Refs #...." to the message and also explain that "A future commit will ...".

**Comment by timgraham** [comment]
Add docs for the new Meta option in docs/ref/models/options.txt? (maybe elsewhere too)

**Comment by akki** [comment]
Updated docs - options.txt and indexest.txt

File: django/db/migrations/autodetector.py [hunk 2]
```diff
@@ -521,6 +522,9 @@ def generate_created_models(self):
                     related_fields[field.name] = field
                 if getattr(field.remote_field, "through", None) and not field.remote_field.through._meta.auto_created:
                     related_fields[field.name] = field
+            # Are there any indexes to defer?
+            indexes = model_state.options['indexes']
```
Comment (timgraham) [inline django/db/migrations/autodetector.py]: Does `.pop()` not work as done below?

File: django/db/backends/base/schema.py [hunk 3]
```diff
@@ -890,6 +890,9 @@ def _model_indexes_sql(self, model):
         for field_names in model._meta.index_together:
             fields = [model._meta.get_field(field) for field in field_names]
             output.append(self._create_index_sql(model, fields, suffix="_idx"))
+
+        for index in model._meta.indexes:
+            output.append(index.create_sql(self))
```
Comment (timgraham) [inline django/db/backends/base/schema.py]: Rebase and add `model` argument.

File: django/db/migrations/state.py [hunk 4]
```diff
@@ -403,6 +407,11 @@ def from_model(cls, model, exclude_rels=False):
                 elif name == "index_together":
                     it = model._meta.original_attrs["index_together"]
                     options[name] = set(normalize_together(it))
+                elif name == "indexes":
+                    indexes = model._meta.original_attrs["indexes"]
+                    for index in indexes:
+                        index.name  # Cache the property
```
Comment (timgraham) [inline django/db/migrations/state.py]: I don't understand why this is needed.

File: docs/ref/models/indexes.txt [hunk 5]
```diff
@@ -8,7 +8,8 @@ Model index reference
 
 .. versionadded:: 1.11
 
-Index classes ease creating database indexes. This document explains the API
+Index classes ease creating database indexes. They can be added using model's
+:ref:`metadata options <meta-options>`. This document explains the API
```
Comment (timgraham) [inline docs/ref/models/indexes.txt]: link to the specifc Meta option rather than the "meta-options" overview

File: django/db/migrations/autodetector.py (hunk 2 above)
Comment (akki) [inline django/db/migrations/autodetector.py]: `.pop()` deletes the key from the dictionary. Unlike other options below, we don't want to delete the key (we want to empty the list).

File: django/db/migrations/state.py (hunk 4 above)
Comment (akki) [inline django/db/migrations/state.py]: Else the [sanity-check](https://github.com/django/django/pull/6857/files#diff-c8a3e1d248f8f5b10af7ac5b1d8479f4R356) would fail for indexes defined without `name` from `Meta`.

**Comment by akki** [comment]
@timgraham @MarkusH Please let me know if anything else is required.

File: django/db/migrations/autodetector.py [hunk 6]
```diff
@@ -581,6 +585,15 @@ def generate_created_models(self):
...
             ]
             related_dependencies.append((app_label, model_name, None, True))
+            for index in indexes:
+                self.add_operation(
+                    app_label,
+                    operations.AddIndex(
+                        model_name=model_name,
+                        index=index,
+                    ),
+                    dependencies=related_dependencies,
```
Comment (timgraham) [inline django/db/migrations/autodetector.py]: No tests fail with this line commented out, can you test it?

File: django/db/migrations/state.py [hunk 7]
```diff
@@ -353,6 +353,10 @@ def __init__(self, app_label, name, fields, options=None, bases=None, managers=N
                     'ModelState.fields cannot refer to a model class - "%s.through" does. '
                     'Use a string reference instead.' % name
                 )
+        # Sanity-check that indexes have their names set.
+        for index in self.options['indexes']:
+            if not index._name:
+                raise ValueError("Indexes passed to ModelState require a name attribute, %r doesn't have one." % index)
```
Comment (timgraham) [inline django/db/migrations/state.py]: Use:
```python
raise ValueError(
    'Indexes passed to ModelState require a name attribute, '
    '%r doesn't have one.' % index
)
```
While we allow longer lines, we typically break up long strings like this.

File: django/db/migrations/state.py (hunk 4 above)
Comment (timgraham) [inline django/db/migrations/state.py]: The design of this doesn't seem right. Contrary to the comment, the purpose seems to be that you need to set `index._name` which accessing `index.name` happens to do. That's not really intuitive. If this can't be remedied right now, the comment should be updated to reflect the true purpose.

File: docs/ref/models/indexes.txt [hunk 8]
```diff
@@ -8,8 +8,9 @@ Model index reference
 
 .. versionadded:: 1.11
 
-Index classes ease creating database indexes. This document explains the API
-references of :class:`Index` which includes the `index options`_.
+Index classes ease creating database indexes. They can be added using the
+:attr:`~django.db.models.Options.indexes` model option. This document explains
```
Comment (timgraham) [inline docs/ref/models/indexes.txt]: using the Meta.indexes option

File: docs/ref/models/options.txt [hunk 9]
```diff
@@ -387,6 +387,28 @@ The name of the manager to use for the model's
...
+.. attribute:: Options.indexes
+
+.. versionadded:: 1.11
+
+    A list of :doc:`indexes </ref/models/indexes>` that you want to define on
+    the model::
+
+        from django.db import models
+
+        class FooBar(models.Model):
```
Comment (timgraham) [inline docs/ref/models/options.txt]: If you could use some real model that's used elsewhere in the docs rather than a meaningless "FooBar", I think that would be an improvement.

File: docs/ref/models/options.txt [hunk 10]
```diff
@@ -387,6 +387,28 @@ The name of the manager to use for the model's
...
+            foo = models.CharField(max_length=100)
+            bar = models.IntegerField()
+
+            class Meta:
+                indexes = [
+                    models.Index(fields=['foo', 'bar']),
+                    models.Index(fields=['foo'], name='my_foo_index')
```
Comment (timgraham) [inline docs/ref/models/options.txt]: add trailing comma

File: tests/migrations/test_autodetector.py [hunk 11]
```diff
@@ -1130,6 +1144,51 @@ def test(from_state, to_state, msg):
         for t in tests:
             test(*t)
 
+    def test_create_model_with_indexes(self):
+        """ Test creation of new model with indexes already defined """
```
Comment (timgraham) [inline tests/migrations/test_autodetector.py]: chop spaces next to """ for consistent with other docstrings

File: tests/migrations/test_autodetector.py [hunk 12]
```diff
@@ -1130,6 +1144,51 @@ def test(from_state, to_state, msg):
...
+    def test_remove_indexes(self):
+        """Test change detection of removed indexes."""
+        changes = self.get_changes([self.author_empty, self.book_indexes], [self.author_empty, self.book])
+        # Right number/type of migrations?
+        self.assertNumberMigrations(changes, 'otherapp', 1)
+        self.assertOperationTypes(changes, 'otherapp', 0, ['RemoveIndex'])
+        self.assertOperationAttributes(changes, 'otherapp', 0, 0, model_name='book', name='book_title_author_idx')
+
+    def test_order_fields_indexes(self):
+        """Test change detection of re-ordering of fields in indexes."""
```
Comment (timgraham) [inline tests/migrations/test_autodetector.py]: reordering (chop dash)

**Comment by timgraham** [comment]
Is adding an index using AddIndex if it's not in Model.indexes (as suggested in migration-operations.txt) valid? Is there a risk Django might drop that index in some future alter operation since it doesn't know about it in models?

**Comment by akki** [comment]
Ahh, right! I think we can remove that from the docs, none of the other operations does that either. I have made this change in the last commit, if it seems right.

I also wanted to know if (and how) I should mention this in the release docs.

Thanks

File: docs/ref/migration-operations.txt [hunk 13]
```diff
@@ -202,21 +202,7 @@ is set, its column name).
...
-        operations = [
-            migrations.AddIndex(
-                'Book',
-                models.Index(fields=['title', 'author'], name='my_index_name'),
-            ),
-        ]
-
-If you're writing your own migration to add an index, you must assign a
-``name`` to the ``index`` as done above.
+It only accepts an `Index` instance for which the `name` attribute is defined.
```
Comment (timgraham) [inline docs/ref/migration-operations.txt]: Not sure this is important to mention since most users shouldn't encounter it and a helpful message is raised in that case. Otherwise, you need to use double backticks around "Index" and "name" to have an effect on the rendered docss.

**Comment by timgraham** [comment]
I think we can wait until the end of this project before writing release notes as things may evolve.

File: docs/ref/migration-operations.txt (hunk 13 above)
Comment (akki) [inline docs/ref/migration-operations.txt]: Better to remove it then.

File: django/db/migrations/autodetector.py [hunk 14]
```diff
@@ -176,6 +176,9 @@ def _detect_changes(self, convert_apps=None, graph=None):
         self.generate_altered_options()
         self.generate_altered_managers()
 
+        # Create the altered indexes and store them in self.altered_indexes
```
Comment (timgraham) [inline django/db/migrations/autodetector.py]: add period

File: django/db/migrations/autodetector.py [hunk 15]
```diff
@@ -176,6 +176,9 @@ def _detect_changes(self, convert_apps=None, graph=None):
...
+        # This is done to avoid same computation in generate_removed_indexes
```
Comment (timgraham) [inline django/db/migrations/autodetector.py]: "This avoids the same computation...."

File: django/db/migrations/autodetector.py [hunk 16]
```diff
@@ -176,6 +176,9 @@ def _detect_changes(self, convert_apps=None, graph=None):
...
+        # and generate_added_indexes afterwards
```
Comment (timgraham) [inline django/db/migrations/autodetector.py]: chop afterwards, add period

File: django/db/backends/sqlite3/schema.py [hunk 17]
```diff
@@ -154,12 +154,20 @@ def is_self_referential(f):
             for index in model._meta.index_together
         ]
 
+        indexes = model._meta.indexes
+        if delete_fields:
+            indexes = [
+                index for index in indexes
+                if delete_fields[0].column not in index.fields
```
Comment (timgraham) [inline django/db/backends/sqlite3/schema.py]: Isn't this comparing column names (`.column`) to field names in `index.fields`? I think it should be `delete_fields[0].name` instead.

File: django/db/migrations/state.py [hunk 18]
```diff
@@ -353,6 +353,13 @@ def __init__(self, app_label, name, fields, options=None, bases=None, managers=N
                     'ModelState.fields cannot refer to a model class - "%s.through" does. '
                     'Use a string reference instead.' % name
                 )
+        # Sanity-check that indexes have their names set.
```
Comment (timgraham) [inline django/db/migrations/state.py]: names -> name

File: django/db/migrations/state.py [hunk 19]
```diff
@@ -403,6 +410,12 @@ def from_model(cls, model, exclude_rels=False):
                 elif name == "index_together":
                     it = model._meta.original_attrs["index_together"]
                     options[name] = set(normalize_together(it))
+                elif name == "indexes":
+                    indexes = model._meta.original_attrs["indexes"]
+                    for index in indexes:
+                        # Set index._name of all indexes by calling index.name
+                        index.name
```
Comment (timgraham) [inline django/db/migrations/state.py]: @MarkusH, wanted to flag this for your review. It doesn't look good to have attribute access have the side effect of setting some other attribute on the class. Agreed?

@akki, Remind me why the name computation can't be done in `Index.__init__()`?

File: docs/ref/models/indexes.txt [hunk 20]
```diff
@@ -8,8 +8,9 @@ Model index reference
 
 .. versionadded:: 1.11
 
-Index classes ease creating database indexes. This document explains the API
-references of :class:`Index` which includes the `index options`_.
+Index classes ease creating database indexes. They can be added using the
+Meta.indexes option. This document explains the API references of :class:`Index`
```
Comment (timgraham) [inline docs/ref/models/indexes.txt]: Meta.Indexes should be linked as it was before

File: docs/ref/models/options.txt [hunk 21]
```diff
@@ -387,6 +387,28 @@ The name of the manager to use for the model's
     See :meth:`django.db.models.Model.save()` for more about the old and
     new saving algorithm.
 
+``indexes``
+-----------
+
+.. attribute:: Options.indexes
+
+.. versionadded:: 1.11
```
Comment (timgraham) [inline docs/ref/models/options.txt]: indent this line

File: django/db/migrations/state.py (hunk 19 above)
Comment (akki) [inline django/db/migrations/state.py]: Because the `Index` class doesn't has it's `model` attribute set in `__init__` which is required in `get_name`.

Comment (timgraham) [inline django/db/migrations/state.py]: A possible idea is to add a method like `Index.set_model()` and use that instead of assigning the `model` attribute directly. Could something like also set the index name?

Comment (akki) [inline django/db/migrations/state.py]: What about a `Index.set_name_using_model(self, model)` method, we don't even need to set a `model` attribute ever for any of the indexes. Setting `_name` in `Index.set_model()` seems equally counter-intuitive to me.

Comment (timgraham) [inline django/db/migrations/state.py]: Maybe with "with_model" rather than "using_model" would be a bit shorter? Otherwise, seems okay to me.

Comment (akki) [inline django/db/migrations/state.py]: Done as the first commit of this PR and rebased(modified) other commits on top of it. Thanks.

File: django/db/migrations/autodetector.py [hunk 22]
```diff
@@ -521,6 +529,9 @@ def generate_created_models(self):
                     related_fields[field.name] = field
                 if getattr(field.remote_field, "through", None) and not field.remote_field.through._meta.auto_created:
                     related_fields[field.name] = field
+            # Are there any indexes to defer?
```
Comment (MarkusH) [inline django/db/migrations/autodetector.py]: This comment doesn't really make sense. Could you explain why you're doing what you're doing?

Comment (akki) [inline django/db/migrations/autodetector.py]: If we have indexes defined in `Meta.indexes` while creating a new model we also create `AddIndex` operations in the migration. Similar to what has been done for `AlterUniqu/IndexTogether` and `order_with_respect_to` [here](https://github.com/django/django/pull/6857/files/2496a55836b0bc941356bf04c57d31e5ecc52d6c#diff-a71b173d008afea1acd5318ef509bb7cR535).

File: django/db/migrations/autodetector.py [hunk 23]
```diff
@@ -919,6 +939,46 @@ def generate_altered_fields(self):
...
                     self._generate_added_field(app_label, model_name, field_name)
 
+    def create_altered_indexes(self):
+        option_name = 'indexes'
+        for app_label, model_name in sorted(self.kept_model_keys):
+            old_model_name = self.renamed_models.get((app_label, model_name), model_name)
+            old_model_state = self.from_state.models[app_label, old_model_name]
+            new_model_state = self.to_state.models[app_label, model_name]
+
+            old_indexes = old_model_state.options.get(option_name)
```
Comment (MarkusH) [inline django/db/migrations/autodetector.py]: Can you use `['indexes']` here? If not, the list comprehensions in the next lines have a `not in None` and will fail.

File: django/db/migrations/autodetector.py [hunk 24]
```diff
@@ -919,6 +939,46 @@ def generate_altered_fields(self):
...
+            rem_idx = [idx for idx in old_indexes if idx not in new_indexes]
+
+            self.altered_indexes.update({
+                (app_label, model_name): {
+                    'added_indexes': add_idx, 'removed_indexes': rem_idx,
+                }
+            })
+
+    def generate_added_indexes(self):
+        for (app_label, model_name), alt_indexes in list(self.altered_indexes.items()):
```
Comment (MarkusH) [inline django/db/migrations/autodetector.py]: no need for a `list()` casting.

File: django/db/migrations/autodetector.py [hunk 25]
```diff
@@ -919,6 +939,46 @@ def generate_altered_fields(self):
...
+                self.add_operation(
+                    app_label,
+                    operations.AddIndex(
+                        model_name=model_name,
+                        index=index,
+                    )
+                )
+
+    def generate_removed_indexes(self):
+        for (app_label, model_name), alt_indexes in list(self.altered_indexes.items()):
```
Comment (MarkusH) [inline django/db/migrations/autodetector.py]: No `list()` casting needed here.

File: django/db/migrations/operations/models.py [hunk 26]
```diff
@@ -759,7 +759,7 @@ class AddIndex(IndexOperation):
 
     def __init__(self, model_name, index):
         self.model_name = model_name
-        if not index._name:
+        if not index.name:
```
Comment (MarkusH) [inline django/db/migrations/operations/models.py]: Why this change again?

File: django/db/migrations/state.py [hunk 27]
```diff
@@ -353,6 +353,13 @@ def __init__(self, app_label, name, fields, options=None, bases=None, managers=N
                     'ModelState.fields cannot refer to a model class - "%s.through" does. '
                     'Use a string reference instead.' % name
                 )
+        # Sanity-check that indexes have their name set.
+        for index in self.options['indexes']:
+            if not index.name:
```
Comment (MarkusH) [inline django/db/migrations/state.py]: `_name`

File: django/db/migrations/operations/models.py (hunk 26 above)
Comment (MarkusH) [inline django/db/migrations/operations/models.py]: Ah, never mind, didn't see the related patch for `set_name_with_model()`.

File: docs/ref/migration-operations.txt [hunk 28]
```diff
@@ -202,22 +202,6 @@ is set, its column name).
 Creates an index in the database table for the model with ``model_name``.
 ``index`` is an instance of the :class:`~django.db.models.Index` class.
 
-For example, to add an index on the ``title`` and ``author`` fields of the
```
Comment (MarkusH) [inline docs/ref/migration-operations.txt]: Why remove this section?

File: django/db/migrations/autodetector.py (hunk 24 above)
Comment (akki) [inline django/db/migrations/autodetector.py]: Nice catch! Can't remember why I added them in the first place.

File: docs/ref/migration-operations.txt (hunk 28 above)
Comment (akki) [inline docs/ref/migration-operations.txt]: Related discussion:
https://github.com/django/django/pull/6857#issuecomment-232463186; https://github.com/django/django/pull/6857#issuecomment-232654484

File: django/db/migrations/autodetector.py [hunk 29]
```diff
@@ -919,6 +939,46 @@ def generate_altered_fields(self):
...
 
+    def create_altered_indexes(self):
+        option_name = operations.AddIndex.option_name
+        for app_label, model_name in sorted(self.kept_model_keys):
+            old_model_name = self.renamed_models.get((app_label, model_name), model_name)
+            old_model_state = self.from_state.models[app_label, old_model_name]
+            new_model_state = self.to_state.models[app_label, model_name]
+
+            old_indexes = old_model_state.options[option_name]
+            new_indexes = new_model_state.options[option_name]
```
Comment (akki) [inline django/db/migrations/autodetector.py]: @MarkusH Just wanted to confirm if this is what you meant [here](https://github.com/django/django/pull/6857#discussion_r71702557) and `new_indexes` [cannot be](https://github.com/django/django/blob/283b468462a586fd9bf7a2794e9b9a20a7e8a2d9/django/db/migrations/state.py#L333) `None`.

Comment (timgraham) [inline django/db/migrations/autodetector.py]: That looks correct to me.

File: django/db/migrations/autodetector.py [hunk 30]
```diff
@@ -175,13 +176,20 @@ def _detect_changes(self, convert_apps=None, graph=None):
         self.generate_altered_options()
         self.generate_altered_managers()
 
+        # Create the altered indexes and store them in self.altered_indexes.
+        # This avoids the same computation in generate_removed_indexes
```
Comment (timgraham) [inline django/db/migrations/autodetector.py]: generate_removed_indexes() and generate_added_indexes()
(add parentheses)

File: django/db/models/base.py [hunk 31]
```diff
@@ -298,6 +298,12 @@ def __new__(cls, name, bases, attrs):
                 else:
                     new_class.add_to_class(field.name, copy.deepcopy(field))
 
+        # Set name for _meta.indexes. We cannot do this in Options.contribute_to_class
```
Comment (timgraham) [inline django/db/models/base.py]: Set the name of _meta.indexes. This can't be done in Options.contribute_to_class() because fields haven't been added to the model yet.

File: django/db/migrations/state.py [hunk 32]
```diff
@@ -353,6 +353,13 @@ def __init__(self, app_label, name, fields, options=None, bases=None, managers=N
...
+                raise ValueError(
+                    "Indexes passed to ModelState require a name attribute, "
```
Comment (timgraham) [inline django/db/migrations/state.py]: , -> . (for consistency with same message in operations/models.py)

File: docs/ref/models/indexes.txt [hunk 33]
```diff
@@ -8,8 +8,10 @@ Model index reference
 
 .. versionadded:: 1.11
 
-Index classes ease creating database indexes. This document explains the API
-references of :class:`Index` which includes the `index options`_.
+Index classes ease creating database indexes. They can be added using the
+:attr:`Meta.indexes <django.db.models.Options.indexes>` option. This document
+explains the API references of :class:`Index` which includes the
```
Comment (timgraham) [inline docs/ref/models/indexes.txt]: Adjust line break:
```
explains the API references of :class:`Index` which includes the `index
options`_.
```

File: tests/migrations/test_operations.py [hunk 34]
```diff
@@ -1447,6 +1452,44 @@ def test_remove_index(self):
         self.assertEqual(definition[1], [])
         self.assertEqual(definition[2], {'model_name': "Pony", 'name': "pony_test_idx"})
```
Comment (timgraham) [inline tests/migrations/test_operations.py]: It looks like most tests are using comments to separate sections rather than a blank line.

**Comment by timgraham** [comment]
I merged the first commit in b92c6b7d5641e927145318adaaca1156ce26ddd2. In the end, unless you feel differently, I think we can squash the others when merging. Splitting them out definitely made review easier though.

I think this is good to go, @MarkusH are you happy?

**Comment by akki** [comment]
I share the same rationale. Will squash them in my next push.

File: django/db/models/base.py (hunk 31 above)
Comment (akki) [inline django/db/models/base.py]: I would like to stick to "till then" than "yet" since the fields have been initialized now at this point. What I want to convey is that fields haven't been added to the model when `contribute_to_class` is called.

File: tests/migrations/test_operations.py (hunk 34 above)
Comment (akki) [inline tests/migrations/test_operations.py]: This is kind of a separate test case but I have it to this method itself because they share the same initial setup. Hence, added the line break as a demarcation.

File: django/db/models/base.py (hunk 31 above)
Comment (timgraham) [inline django/db/models/base.py]: How about "till then" -> "at that point"?

Comment (akki) [inline django/db/models/base.py]: Great. Done. Thanks!

**Comment by timgraham** [comment]
merged in 6a8372e6ec42eb50e05f9dbcb26b783237bdc233, thanks!
